pkg/kubelet/cm/dra/manager.go

```diff
@@ -108,30 +109,21 @@ func (m *ManagerImpl) PrepareResources(pod *v1.Pod) error {
 			continue
 		}
 
-		claimInfo := m.cache.get(*claimName, pod.Namespace)
-		if claimInfo == nil {
-			// claim does not exist in cache, create new claimInfo object
-			// to be processed later.
-			claimInfo = newClaimInfoFromResourceClaim(resourceClaim)
-		}
-
-		// We delay checkpointing of this change until this call
-		// returns successfully. It is OK to do this because we
-		// will only return successfully from this call if the
-		// checkpoint has succeeded. That means if the kubelet is
-		// ever restarted before this checkpoint succeeds, the pod
-		// whose resources are being prepared would never have
-		// started, so it's OK (actually correct) to not include it
-		// in the cache.
-		claimInfo.addPodReference(pod.UID)
-
-		if claimInfo.prepared {
-			// Already prepared this claim, no need to prepare it again
-			continue
+		claimInfo, ok := m.cache.get(*claimName, pod.Namespace)
+		if ok {
+			// Claim exists in cache. Reference the pod.
+			m.cache.addPodReference(*claimName, pod.Namespace, pod.UID)
+			if claimInfo.prepared {
+				// Already prepared this claim, no need to prepare it again
+				continue
+			}
 		}
+		// This saved claim will be used to update ClaimInfo cache
+		// after NodePrepareResources GRPC succeeds
+		resourceClaims[resourceClaim.UID] = resourceClaim
 
 		// Loop through all plugins and prepare for calling NodePrepareResources.
-		for _, resourceHandle := range claimInfo.ResourceHandles {
+		for _, resourceHandle := range resourceClaim.Status.Allocation.ResourceHandles {
```

Comment:
It feels like this could be done more cleanly. The previous code was very clear (though buggy) -- looking at this new code, it's not as obvious at first glance. Need to think about how to improve it.

Reply:
I tend to disagree here. I think this is much more clear. claimInfo.ResourceHandles was hiding the source of data, but this code shows it explicitly and clearly.

Reply:
This comment is about this whole code block (I highlighted lines 111-126), not resourceHandles specifically.

Reply:
ah, ok. Thanks for the clarification. How about this one that I posted in another discussion?
```
claimInfo, ok := m.cache.get(*claimName, pod.Namespace)
if ok {
        m.cache.addPodReference(*claimName, pod.Namespace, pod.UID)
}

if claimInfo.prepared {
    continue
}

// I intentionally do it here, after the above check
resourceClaims[resourceClaim.UID] = resourceClaim 
```

It's more readable, but hides that it checks claimInfo.prepared for no reason when claimInfo is not in the cache.

Reply:
Did as suggested. Thanks.